include/dali/core/mm/cuda_vm_resource.h

```diff
@@ -659,14 +665,47 @@ class cuda_vm_resource_base : public memory_resource<memory_kind::device> {
       to_map.push_back({ next_unmapped, block_idx });
     }
 
-    // Prevent `get_free_blocks` from unmapping blocks that are already mapped to this range.
-    mark_as_unavailable(ptr, size);
+    // a list of blocks to unmap
+    SmallVector<std::pair<va_region*, int>, 256> to_unmap;
+    {
+      // Prevent `get_free_blocks` from unmapping blocks that are already mapped to this range.
+      available_backup_ = region->available;
+      int num_avail_backup = region->available_blocks;
+      mark_as_unavailable(ptr, size);
+
+      // Let's find some free blocks which we can reuse.
+      get_free_blocks(to_unmap, blocks_to_map);
 
+      // Restore original availability mask
+      region->available = available_backup_;
+      region->available_blocks = num_avail_backup;
+    }
+
+    int new_blocks = blocks_to_map - to_unmap.size();
     SmallVector<cuvm::CUMem, 256> free_blocks;
-    get_free_blocks(free_blocks, blocks_to_map);
-    assert(static_cast<int>(free_blocks.size()) == blocks_to_map);
+    free_blocks.resize(blocks_to_map);
+    for (int i = to_unmap.size(); i < blocks_to_map; i++) {
+      // This can throw, but we should leave the object in a consistent state
+      free_blocks[i] = cuvm::CUMem::Create(block_size_, device_ordinal_);
+    }
 
     int i = 0;
+    for (auto &region_block : to_unmap) {
+      // Each iteration should leave the resource in a consistent state.
+      va_region *r = region_block.first;
+      int block_idx = region_block.second;
+      auto mem = r->unmap_block(block_idx);  // This can (potentially) throw
+      stat_.total_unmaps++;
+      char *ptr = r->block_ptr<char>(block_idx);
+      free_mapped_.get_specific_block(ptr, block_size_);
+      stat_take_free(block_size_);
+      free_blocks[i++] = std::move(mem);
+    }
+    assert(free_blocks.size() == static_cast<size_t>(blocks_to_map));
+
+    stat_allocate_blocks(new_blocks);
```

Comment:
move this to line 691?

Reply:
Nope - if `unmapping` throws for any reason, the blocks will just go away.